frontends/systolic-lang/gen_post_op.py

```diff
@@ -0,0 +1,329 @@
+        fp_mult.left = numeric_types.FixedPoint(
+            str(float_to_fixed_point(0.01, FRACWIDTH)), BITWIDTH, INTWIDTH, True
+        ).unsigned_integer()
+        fp_mult.right = this.value
+        fp_mult.go = ~(write_mem.out) @ 1
+
+        # Increment idx_reg during the cycle that we write to memory.
+        incr_idx.left = this.idx_reg_out
+        incr_idx.right = 1
+        this.idx_reg_in = write_mem.out @ incr_idx.out
+        this.idx_reg_write_en = write_mem.out @ 1
+
+        # Write to memory.
+        g.asgn(write_en_port, 1, write_mem.out)
+        g.asgn(addr0_port, this.index)
+        g.asgn(write_data_port, this.value, ~lt.out)
+        g.asgn(write_data_port, fp_mult.out, lt.out)
+        # Groups is done once we have written to memory.
+        g.done = write_done_port
+
+    comp.control = py_ast.Enable("do_relu")
+
+
+def create_leaky_relu_groups(comp: cb.ComponentBuilder, row, num_cols, addr_width):
+    """
+    Creates the groups for the leaky relu post op, i.e., the post-op that
+    coordinates the leaky-relu execution.
+    """
+
+    def store_output_vals(comp: cb.ComponentBuilder, row, num_cols, addr_width):
+        """
+        Helper function that looks at the systolic array output signsl (e.g.,
+        `r0_valid`, `r0_value`, etc.) and creates signals that tells us when: a)
+        each row is ready for the leaky relu operations to start and b)
+        the output systolic array values (we need them in registers bc the systolic
+        array outputs are only available for one cycle).
+        """
+        this = comp.this()
+        row_ready_wire = comp.wire(f"r{row}_ready_wire", 1)
+        row_ready_reg = comp.reg(f"r{row}_ready_reg", 1)
+        for col in range(num_cols):
+            wire_value = comp.wire(f"r{row}_c{col}_val_wire", BITWIDTH)
+            reg_value = comp.reg(f"r{row}_c{col}_val_reg", BITWIDTH)
+            val_ready = comp.wire(f"r{row}_c{col}_val_ready", 1)
+            valid_signal = this.port(f"r{row}_valid")
+            idx_signal = this.port(f"r{row}_idx")
+            value_signal = this.port(f"r{row}_value")
+            with comp.static_group(f"r{row}_c{col}_value_group", 1):
+                # Wire to detect and hold when the row is first valid. Once
+                # it is valid, we can safely start our relu operations.
+                row_ready_reg.in_ = valid_signal @ 1
+                row_ready_reg.write_en = valid_signal @ 1
+                row_ready_wire.in_ = (valid_signal | row_ready_reg.out) @ 1
+                # Logic to hold the systolic array output values. We need registers
+                # because the output values for systolic arrays are only available
+                # for one cycle before they change.
+                val_ready.in_ = (
+                    valid_signal
+                    & (
+                        idx_signal
+                        == cb.ExprBuilder(py_ast.ConstantPort(addr_width, col))
```

Comment:
Can we hoist this out into a variable? It's kind of hard to read and doesn't seem like it depends on the assignments in the group anyways.